File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let carry = 0, head = null, curr = null;

    while (l1 || l2) {
        let v1 = 0, v2 = 0
        if (l1) v1 = l1.val;
        if (l2) v2 = l2.val;

        let newVal = v1 + v2 + carry;
        let newCarry = Math.floor(newVal / 10);
        newVal %= 10;
        let nde = new ListNode(newVal);
        if (!head) {
            head = nde;
        } else {
            curr.next = nde;
        }
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
        curr = nde;
        carry = newCarry;
    }
    if (carry !== 0) {
        curr.next = new ListNode(carry);
    }
    return head;
};

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode();
        ListNode prev = null;
        int carry = 0;

        while (l1 != null || l2 != null) {
            int val1 = 0; int val2 = 0;

            if (l1 != null) {
                val1 = l1.val;
                l1 = l1.next;
            } 
            if (l2 != null) {
                val2 = l2.val;
                l2 = l2.next;
            }
            int val = val1 + val2 + carry;
            carry = (int) Math.floor(val / 10);
            ListNode nde = new ListNode(val % 10);
            if (prev == null) {
                result = nde;
                prev = nde;
            } else {
                prev.next = nde;
                prev = nde;
            }
            if (l1 == null && l2 == null && carry > 0) {
                prev.next = new ListNode(carry);
            }
        }
        return result;
    }
}